File: moseoh/_8주차/오픈채팅방/java/Solution.java
package _8주차.오픈채팅방.java;

import java.util.*;

class Solution {
    public static void main(String[] args) {
        String[] record1 = new String[]{"Enter uid1234 Muzi", "Enter uid4567 Prodo", "Leave uid1234", "Enter uid1234 Prodo", "Change uid4567 Ryan"};
        String[] answer1 = new String[]{"Prodo님이 들어왔습니다.", "Ryan님이 들어왔습니다.", "Prodo님이 나갔습니다.", "Prodo님이 들어왔습니다."};
        String[] result1 = new Solution().solution(record1);
        System.out.printf(
                "테스트 케이스 1: %s\n" +
                        "\t- 실행 결과: %s\n" +
                        "\t- 기댓값: %s\n\n",
                Arrays.equals(result1, answer1) ? "정답" : "**오답**", Arrays.toString(result1), Arrays.toString(answer1)
        );
    }

    public String[] solution(String[] record) {
        List<String> recordList = new ArrayList<>();
        Map<String, String> nameMap = new HashMap<>();
        for (String s : record) {
            if (!s.startsWith("Leave")) {
                String[] split = s.split(" ");
                nameMap.put(split[1], split[2]);
            }

            if (!s.startsWith("Change")) {
                recordList.add(s);
            }
        }

        int length = recordList.size();
        String[] answer = new String[length];
        for (int i = 0; i < length; i++) {
            String[] split = recordList.get(i).split(" ");
            answer[i] = nameMap.get(split[1]);
            if ("Enter".equals(split[0])) {
                answer[i] += "님이 들어왔습니다.";
            } else {
                answer[i] += "님이 나갔습니다.";
            }
        }

        return answer;
    }
}

File: moseoh/_8주차/오픈채팅방/kotlin/Solution.kt
package _8주차.오픈채팅방.kotlin

fun main() {
    val record1 = arrayOf(
        "Enter uid1234 Muzi",
        "Enter uid4567 Prodo",
        "Leave uid1234",
        "Enter uid1234 Prodo",
        "Change uid4567 Ryan"
    )
    val answer1 = arrayOf("Prodo님이 들어왔습니다.", "Ryan님이 들어왔습니다.", "Prodo님이 나갔습니다.", "Prodo님이 들어왔습니다.")
    val result1 = Solution().solution(record1)
    check(result1.contentEquals(answer1)) {
        "\n\n테스트 케이스 1\n" +
                "\t- 실행 결과: ${result1.contentToString()}\n" +
                "\t- 기댓값: ${answer1.contentToString()}\n"
    }
}

class Solution {
    fun solution(record: Array<String>): Array<String> {
        val users = hashMapOf<String, String>()
        val accessRecord = mutableListOf<Pair<Action, String>>()

        record.forEach {
            val split = it.split(" ")
            when (Action.valueOf(split[0])) {
                Action.Change -> {
                    val uuid = split[1]
                    val nickname = split[2]
                    users[uuid] = nickname
                }

                Action.Enter -> {
                    val uuid = split[1]
                    val nickname = split[2]
                    users[uuid] = nickname

                    accessRecord.add(Pair(Action.Enter, uuid))
                }

                Action.Leave -> {
                    val uuid = split[1]

                    accessRecord.add(Pair(Action.Leave, uuid))
                }
            }
        }

        return accessRecord.map {
            users[it.second] + it.first.message
        }.toTypedArray()
    }

    enum class Action(val message: String) {
        Enter("님이 들어왔습니다."),
        Leave("님이 나갔습니다."),
        Change("");
    }
}